internal/addonsdk/sdk.go

```diff
@@ -0,0 +1,192 @@
+				},
+			},
+			stopperCh: make(chan bool),
+			updateCh:  make(chan updateOptions),
+			doneCh:    make(chan bool),
+			log:       logger,
+		}
+		// because the heartbeat reporter still hasn't been started
+		defer close(statusReporterSingleton.doneCh)
+	})
+	return statusReporterSingleton
+}
+
+func (sr *statusReporter) Start(ctx context.Context) error {
+	// ensures to tie only one heartbeat-reporter loop at a time to a StatusReporter object
+	select {
+	case <-sr.doneCh:
+		sr.log.Info("StatusReporter already found to be running. Please explicitly Stop() it first before trying to restart it")
+		return nil
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+	}
+	sr.doneCh = make(chan bool)
+	defer close(sr.doneCh)
+
+	currentAddonInstance := &addonsv1alpha1.AddonInstance{}
+	if err := sr.addonInstanceInteractor.GetAddonInstance(ctx, types.NamespacedName{Name: "addon-instance", Namespace: sr.addonTargetNamespace}, currentAddonInstance); err != nil {
+		return fmt.Errorf("error occurred while fetching the current heartbeat update period interval: %w", err)
+	}
+	sr.interval = currentAddonInstance.Spec.HeartbeatUpdatePeriod.Duration
+	defer sr.ticker.Stop()
+	sr.ticker = time.NewTicker(sr.interval)
+
+	for {
+		select {
+		case update := <-sr.updateCh:
+			// update the interval if the newInterval in the `update` is provided and is not equal to the existing interval
+			// synchronize the timer with this new interval
+			if update.addonInstance != nil {
+				if update.addonInstance.Spec.HeartbeatUpdatePeriod.Duration != sr.interval {
+					sr.interval = update.addonInstance.Spec.HeartbeatUpdatePeriod.Duration
+					sr.ticker.Reset(sr.interval)
+				}
+			}
+
+			if update.conditions != nil {
+				sr.latestConditions = *update.conditions
+			}
+		case <-sr.ticker.C:
+			if err := sr.updateAddonInstanceStatus(ctx, sr.latestConditions); err != nil {
+				sr.log.Error(err, "failed to report the regular heartbeat")
+			}
+		case <-ctx.Done():
+			return nil
+		case <-sr.stopperCh:
+			return nil
+		}
+	}
+}
+
+func (sr *statusReporter) Stop(ctx context.Context) error {
+	select {
+	case <-sr.doneCh: // will non-blockingly receive whenever doneCh would be closed
+		sr.log.Info("status reporter is already stopped")
+		return nil
+	case sr.stopperCh <- true:
+		return nil
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+func (sr *statusReporter) SetConditions(ctx context.Context, conditions []metav1.Condition) error {
+	// immediately register a new heartbeat upon receive one from the client/tenant side
+	addonInstance := &addonsv1alpha1.AddonInstance{}
+	if err := sr.addonInstanceInteractor.GetAddonInstance(ctx, types.NamespacedName{Name: "addon-instance", Namespace: sr.addonTargetNamespace}, addonInstance); err != nil {
+		return fmt.Errorf("failed to get the AddonInstance: %w", err)
+	}
+
+	// making a deep-copy for current Conditions for rolling back in case of failures
+	previousConditions := addonInstance.Status.Conditions
```

Comment:
This is NOT creating a copy of the slice and it's also not creating a copy of the items within the array.
A slice is just a pointer, so you just assign the same pointer to a new variable.

Reply:
Ahh silly mistake, thanks for catching this!